src/modules/job-manager/jobtap.c

```diff
@@ -0,0 +1,520 @@
+             */
+            if (job->state == FLUX_JOB_STATE_SCHED)
+                flux_log (jobtap->ctx->h, LOG_ERR,
+                          "jobtap: %s: %ju: BUG: plugin didn't return priority",
+                          jobtap_plugin_name (jobtap->plugin),
+                          (uintmax_t) job->id);
+        }
+        /*
+         *   O/w, plugin provided a new priority.
+         */
+    }
+    else if (rc == 0) {
+        /*
+         *  No priority.get callback was run. Enable default behavior
+         *   (priority == urgency)
+         */
+        priority = job->urgency;
+    }
+    else {
+        /*
+         *  priority.get callback was run and failed. Log the error
+         *   and return the current priority.
+         */
+        flux_log (jobtap->ctx->h, LOG_ERR,
+                  "jobtap: %s: job.priority.get: callback failed",
+                  jobtap_plugin_name (jobtap->plugin));
+        priority = job->priority;
+    }
+
+    flux_plugin_arg_destroy (args);
+    *pprio = priority;
+    return rc;
+}
+
+int jobtap_call (struct jobtap *jobtap,
+                 struct job *job,
+                 const char *topic,
+                 const char *fmt,
+                 ...)
+{
+    int rc = -1;
+    json_t *note = NULL;
+    flux_plugin_arg_t *args;
+    int64_t priority = -1;
+    va_list ap;
+
+    if (!jobtap->plugin) {
+        /*
+         * Default with no plugin: ensure we advance past PRIORITY state
+         */
+        if (job->state == FLUX_JOB_STATE_PRIORITY)
+            reprioritize_job (jobtap->ctx, job, job->urgency);
+        return 0;
+    }
+
+    va_start (ap, fmt);
+    if (!(args = jobtap_args_vcreate (jobtap, job, fmt, ap))) {
+        flux_log (jobtap->ctx->h, LOG_ERR,
+                  "jobtap: %s: %s: %ju: failed to create plugin args",
+                  jobtap_plugin_name (jobtap->plugin),
+                  topic,
+                  (uintmax_t) job->id);
+        return -1;
+    }
+    va_end (ap);
+
+    rc = flux_plugin_call (jobtap->plugin, topic, args);
+    if (rc < 0) {
+        flux_log (jobtap->ctx->h, LOG_ERR,
+                  "jobtap: %s: %s: callback returned error",
+                  jobtap_plugin_name (jobtap->plugin),
+                  topic);
+    }
+    if (flux_plugin_arg_unpack (args, FLUX_PLUGIN_ARG_OUT,
+                                "{s?I s?o}",
+                                "priority", &priority,
+                                "annotations", &note) < 0) {
+        flux_log (jobtap->ctx->h, LOG_ERR,
+                  "jobtap: %s: %s: arg_unpack: %s",
+                  jobtap_plugin_name (jobtap->plugin),
+                  topic,
+                  flux_plugin_arg_strerror (args));
+        rc = -1;
+    }
+    if (note && strcmp (topic, "job.new") != 0) {
+        /*
+         *  Allow plugins to update annotations (except in job.new callback)
+         *   (A failure here will be logged but not considered a fatal error)
```

Comment:
Adding a special annotation for jobtap plugins might be wise down the road. Let's not go there now since I'm unsure how this will be used. Eventually there may be a stack of plugins loaded at once, and I'd hate to reduce flexibility by setting something in stone now.

> why not with job.new?

Ah, there should be a comment here. If you set a job annotation before the first journal event, you get lots of complaints from job-info. Easier to just ignore annotations in this callback.

Reply:
Actually, annotations could probably be added in job.new if we just updated the annotations and didn't publish the result. I wonder if that is useful.